plugin/secrets_service_account_key.go

```diff
@@ -213,13 +297,90 @@ func (b *backend) getSecretKey(ctx context.Context, s logical.Storage, rs *RoleS
 	}
 
 	resp := b.Secret(SecretTypeKey).Response(secretD, internalD)
-	resp.Secret.Renewable = true
+	resp.Secret.Renewable = false
+	resp.Secret.MaxTTL = cfg.MaxTTL
+	resp.Secret.TTL = ttlToUse
+
+	return resp, nil
+}
+
+func (b *backend) constructRespFromCache(item *serviceAccountKeyCacheItem, ttl, maxTTL time.Duration) *logical.Response {
+	secretD, internalD := item.secretResponse()
+	resp := b.Secret(SecretTypeKey).Response(secretD, internalD)
+	resp.Secret.TTL = ttl
+	resp.Secret.MaxTTL = maxTTL
+	resp.Secret.Renewable = false
 
-	if ttl > 0 {
-		resp.Secret.TTL = time.Duration(ttl) * time.Second
+	return resp
+}
+
+func getSecretKeyFromCache(ctx context.Context, s logical.Storage, rs *RoleSet) (*serviceAccountKeyCacheItem, error) {
+	cacheCollection, err := getCacheCollection(ctx, s, rs.Name)
+	if err != nil {
+		return nil, errwrap.Wrapf("could not get service account key cache collection: {{err}}", err)
 	}
 
-	return resp, nil
+	if cacheCollection != nil {
+		_, validCacheItem := cacheCollection.getFirstValidItem(rs.bindingHash())
+		if validCacheItem != nil {
+			validCacheItem.Counter++
+
+			if err := cacheCollection.putToStorage(ctx, s, rs.Name); err != nil {
+				return nil, errwrap.Wrapf("failed to update cache key collection: {{err}}", err)
+			}
+
+			return validCacheItem, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func createCacheCollection(ctx context.Context, s logical.Storage, rs *RoleSet, key *iam.ServiceAccountKey, ttl time.Duration) error {
+	now := time.Now()
+
+	cacheTTL := ttl
+	if cacheTTL == 0 {
+		cacheTTL = defaultCacheTTL
+	}
+
+	newCacheItem := &serviceAccountKeyCacheItem{
+		Name:               key.Name,
+		RolesetName:        rs.Name,
+		RolesetBindingHash: rs.bindingHash(),
+		KeyAlgorithm:       key.KeyAlgorithm,
+		PrivateKeyData:     key.PrivateKeyData,
+		KeyType:            key.PrivateKeyType,
+		IssueTime:          now,
+		TTL:                cacheTTL,
+		Counter:            1,
+	}
+
+	newCacheCollection := newServiceAccountKeyCacheCollection(newCacheItem, key.Name)
+
+	if err := newCacheCollection.putToStorage(ctx, s, rs.Name); err != nil {
+		return errwrap.Wrapf("failed to insert new cache collection into storage: {{err}}", err)
+	}
+
+	return nil
+}
+
+func getCacheCollection(ctx context.Context, s logical.Storage, keyName string) (*serviceAccountKeyCacheCollection, error) {
+	cachedKeyCollection, err := s.Get(ctx, keyName)
+	if err != nil {
+		return nil, err
+	}
+
+	if cachedKeyCollection != nil {
+		decodedCollection := new(serviceAccountKeyCacheCollection)
+		if err := cachedKeyCollection.DecodeJSON(decodedCollection); err != nil {
+			return nil, err
+		}
+
+		return decodedCollection, nil
+	}
+
+	return nil, nil
```

Comment:
How about structuring it like this instead?

```
	cachedKeyCollection, err := s.Get(ctx, keyName)
	if err != nil {
		return nil, err
	}

	if cachedKeyCollection == nil {
		return nil, nil
	}

	decodedCollection := new(serviceAccountKeyCacheCollection)
	if err := cachedKeyCollection.DecodeJSON(decodedCollection); err != nil {
		return nil, err
	}

	return decodedCollection, nil
```

Reply:
I think that's better. I'll update this block in the next commit. Thank you.